src/client/Client.cc

```diff
@@ -12896,6 +12912,130 @@ int Client::ll_mknodx(Inode *parent, const char *name, mode_t mode,
   return r;
 }
 
+void Client::_async_create_cb(MetaRequest *req, MetaSession *session, int err)
+{
+  Inode *diri = req->inode();
+  Dentry *dn = req->dentry();
+  unsigned want = CEPH_CAP_FILE_EXCL | CEPH_CAP_DIR_CREATE;
+
+  if (req->async) {
+    InodeRef &in = req->target;
+    if (err < 0) {
+      lderr(cct) << "got error " << cpp_strerror(err) << " for async create tid "
+		 << req->tid << " " << diri->ino  << " " << dn->name << dendl;
+      diri->set_async_err(err);
+      in->set_async_err(err);
+    }
+    put_cap_ref(diri, want);
+    return;
+  }
+
+  Cap *auth_cap = diri->auth_cap;
+  if (!auth_cap || auth_cap->session != session)
+    return;
+  if ((diri->caps_issued() & want) != want)
+    return;
+
+  // We have CEPH_CAP_FILE_EXCL, which implies CEPH_CAP_FILE_SHARED.
+  if (dn->cap_shared_gen != diri->shared_gen ||
+      dn->inode)
+    return;
+
+  if (session->delegated_inos.empty())
+    return;
+
+  if (!diri->async_create_stat)
+    return;
+
+  auto& open_args = req->head.args.open;
+
+  InodeStat st;
+  st.layout = diri->async_create_stat->layout;
+  if (open_args.stripe_unit)
+    st.layout.stripe_unit = open_args.stripe_unit;
+  if (open_args.stripe_count)
+    st.layout.stripe_count = open_args.stripe_count;
+  if (open_args.object_size)
+    st.layout.object_size = open_args.object_size;
+  if ((__s32)open_args.pool >= 0)
+    st.layout.pool_id = open_args.pool;
+  if (!st.layout.is_valid())
+    return;
+
+  st.vino.ino = session->take_delegated_ino();
+  st.vino.snapid = CEPH_NOSNAP;
+  st.version = 1;
+  st.xattr_version = 1;
+
+  st.cap.wanted = ceph_caps_for_mode(ceph_flags_to_mode(open_args.flags));
+  st.cap.caps = st.cap.wanted | CEPH_CAP_ANY_RD;
+  st.cap.realm = diri->snaprealm->ino;
+  st.cap.cap_id = 1;
+  st.cap.seq = 0;
+  st.cap.mseq = 0;
+  st.cap.flags = CEPH_CAP_FLAG_AUTH;
+
+  st.ctime = st.btime = st.mtime = st.atime = req->op_stamp;
+  // only allow writing to the first object before writeable range is persistent
+  st.max_size = st.layout.stripe_unit;
+  st.truncate_size = -1ull;
+  st.truncate_seq = 1;
+  st.mode = open_args.mode | S_IFREG;
+  st.uid = req->head.caller_uid;
+  st.gid = (diri->mode & S_ISGID) ? diri->gid : req->head.caller_gid;
+  st.xattrbl = req->data;
+  st.inline_version = CEPH_INLINE_NONE;
+
+  Inode *in = add_update_inode(&st, req->sent_stamp, session, req->perms);
+  req->target = in;
+  req->created_ino = in->ino;
+
+  diri->mtime = req->op_stamp;
+  diri->dirstat.nfiles++;
+  diri->dir_ordered_count++;
+  clear_dir_complete_and_ordered(diri, false);
+  Dir *dir = diri->open_dir();
+  dn = link(dir, dn->name, in, dn);
+  dn->primary_link = true;
+
+  get_cap_ref(diri, want);
+  req->async = true;
+  req->resend_mds = session->mds_num;
+  uint32_t orig_flags = req->head.args.open.flags;
+  req->head.args.open.flags = orig_flags | CEPH_O_EXCL;
+  register_unsafe_request(req, session);
+  ldout(cct, 10) << __func__ << " succeed" <<  dendl;
+}
```

Comment:
All of the above is done before the request is ever sent. Why? I would think we'd want to send the request as quickly as we can, and then do the above just afterward. If the send fails for some reason, then you'd have less to unwind...